{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab inline\n",
    "import scipy.stats as stats\n",
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "rcParams['font.size']=12\n",
    "rcParams['pdf.fonttype']=42\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load replicates map\n",
    "replicatemap = pd.read_csv('replicate_map.csv',index_col=0,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pDNA -> fill out cell ID\n",
    "replicatemap['DepMap_ID'].fillna('pDNA',inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th>pDNA_batch</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>replicate_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>MIA Paca-2-311Cas9 Rep A p6_batch0</td>\n",
       "      <td>ACH-000601</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MIA Paca-2-311Cas9 Rep B p6_batch0</td>\n",
       "      <td>ACH-000601</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MIA Paca-2-311Cas9 Rep C p6_batch0</td>\n",
       "      <td>ACH-000601</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     DepMap_ID  pDNA_batch\n",
       "replicate_ID                                              \n",
       "MIA Paca-2-311Cas9 Rep A p6_batch0  ACH-000601           0\n",
       "MIA Paca-2-311Cas9 Rep B p6_batch0  ACH-000601           0\n",
       "MIA Paca-2-311Cas9 Rep C p6_batch0  ACH-000601           0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replicatemap.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "564"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# uniquce cell lines\n",
    "len(replicatemap['DepMap_ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pDNA batches\n",
    "replicatemap['pDNA_batch'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ekim8/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: FutureWarning: get_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead\n",
      "  \"\"\"\n",
      "/home/ekim8/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: FutureWarning: get_value is deprecated and will be removed in a future release. Please use .at[] or .iat[] accessors instead\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "# map batch, controls, samples\n",
    "controls = dict()  # [batch] list[repid]\n",
    "samples = dict()  # [batch][cell] list[repid]\n",
    "for repid in replicatemap.index:\n",
    "    cell = replicatemap.get_value(repid,'DepMap_ID')\n",
    "    batch = replicatemap.get_value(repid,'pDNA_batch')\n",
    "    if cell == 'pDNA':\n",
    "        if batch not in controls:\n",
    "            controls[batch] = list()\n",
    "        controls[batch].append(repid)\n",
    "    else:\n",
    "        if batch not in samples:\n",
    "            samples[batch] = dict()\n",
    "        if cell not in samples[batch]:\n",
    "            samples[batch][cell] = list()\n",
    "        samples[batch][cell].append(repid)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: ['pDNA Avana4_010115_1.5Ex_batch1',\n",
       "  'pDNA Avana4_060115_1.5Ex_batch1',\n",
       "  'pDNA Avana4_0101215_0.55Ex_batch1',\n",
       "  'pDNA Avana4_060115_0.55Ex_batch1'],\n",
       " 0: ['pDNA Avana4_010115_1.5Ex_batch0', 'pDNA Avana4_060115_1.5Ex_batch0'],\n",
       " 3: ['Avana 4+ Hu pDNA (M-AA40, 9/30/15)_batch3',\n",
       "  'Avana 4+ Hu pDNA (M-AA40, 9/30/15) (0.2pg/uL)_batch3',\n",
       "  'Avana 4+ Hu pDNA (M-AA40, 9/30/15) 0.2pg/uL_batch3',\n",
       "  'Avana 4+ Hu pDNA (M-AF34, 11/27/18) 0.2pg/uL_batch3'],\n",
       " 2: ['Avana4pDNA20160601-311cas9 RepG09_batch2',\n",
       "  'Avana4pDNA20160601-311cas9 RepG10_batch2',\n",
       "  'Avana4pDNA20160601-311cas9 RepG11_batch2',\n",
       "  'Avana4pDNA20160601-311cas9 RepG12_batch2']}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "controls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load DepMap 19Q2 readcount file\n",
    "readcount = pd.read_csv('raw_readcounts.csv',index_col=0,header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>143B-311Cas9_RepA_p6_batch3</th>\n",
       "      <th>2313287-311Cas9_RepA_p5_batch3</th>\n",
       "      <th>2313287-311Cas9_RepB_p5_batch3</th>\n",
       "      <th>253J-311Cas9_RepA_p5_batch3</th>\n",
       "      <th>42-MG-BA-311Cas9_RepA_p6_batch3</th>\n",
       "      <th>42-MG-BA-311Cas9_RepB_p6_batch3</th>\n",
       "      <th>5637-311Cas9_RepA_p6_batch3</th>\n",
       "      <th>5637-311Cas9_RepB_p6_batch3</th>\n",
       "      <th>59M-311Cas9_RepA_p4_batch3</th>\n",
       "      <th>59M-311Cas9_RepB_p4_batch3</th>\n",
       "      <th>...</th>\n",
       "      <th>YKG1-311Cas9_RepA_p6_batch3</th>\n",
       "      <th>YKG1-311Cas9_RepB_p6_batch3</th>\n",
       "      <th>ZR-75-1-311Cas9_RepA_p5_batch2</th>\n",
       "      <th>ZR-75-1-311Cas9_RepB_p5_batch2</th>\n",
       "      <th>pDNA Avana4_010115_1.5Ex_batch0</th>\n",
       "      <th>pDNA Avana4_010115_1.5Ex_batch1</th>\n",
       "      <th>pDNA Avana4_0101215_0.55Ex_batch1</th>\n",
       "      <th>pDNA Avana4_060115_0.55Ex_batch1</th>\n",
       "      <th>pDNA Avana4_060115_1.5Ex_batch0</th>\n",
       "      <th>pDNA Avana4_060115_1.5Ex_batch1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>833.0</td>\n",
       "      <td>327.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>546.0</td>\n",
       "      <td>213.0</td>\n",
       "      <td>349.0</td>\n",
       "      <td>280.0</td>\n",
       "      <td>489.0</td>\n",
       "      <td>683.0</td>\n",
       "      <td>373.0</td>\n",
       "      <td>...</td>\n",
       "      <td>565.0</td>\n",
       "      <td>556.0</td>\n",
       "      <td>226.0</td>\n",
       "      <td>386.0</td>\n",
       "      <td>305.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>190.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>289.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>1089.0</td>\n",
       "      <td>383.0</td>\n",
       "      <td>391.0</td>\n",
       "      <td>1151.0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>727.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>770.0</td>\n",
       "      <td>1279.0</td>\n",
       "      <td>720.0</td>\n",
       "      <td>...</td>\n",
       "      <td>637.0</td>\n",
       "      <td>773.0</td>\n",
       "      <td>632.0</td>\n",
       "      <td>673.0</td>\n",
       "      <td>398.0</td>\n",
       "      <td>396.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>445.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>95.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>...</td>\n",
       "      <td>171.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>69.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1271 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      143B-311Cas9_RepA_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        833.0   \n",
       "AAAAAACCCGTAGATAGCCT                       1089.0   \n",
       "AAAAAAGAAGAAAAAACCAG                         95.0   \n",
       "\n",
       "                      2313287-311Cas9_RepA_p5_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                           327.0   \n",
       "AAAAAACCCGTAGATAGCCT                           383.0   \n",
       "AAAAAAGAAGAAAAAACCAG                            32.0   \n",
       "\n",
       "                      2313287-311Cas9_RepB_p5_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                           188.0   \n",
       "AAAAAACCCGTAGATAGCCT                           391.0   \n",
       "AAAAAAGAAGAAAAAACCAG                            69.0   \n",
       "\n",
       "                      253J-311Cas9_RepA_p5_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        546.0   \n",
       "AAAAAACCCGTAGATAGCCT                       1151.0   \n",
       "AAAAAAGAAGAAAAAACCAG                         72.0   \n",
       "\n",
       "                      42-MG-BA-311Cas9_RepA_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                            213.0   \n",
       "AAAAAACCCGTAGATAGCCT                            736.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             46.0   \n",
       "\n",
       "                      42-MG-BA-311Cas9_RepB_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                            349.0   \n",
       "AAAAAACCCGTAGATAGCCT                            727.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             60.0   \n",
       "\n",
       "                      5637-311Cas9_RepA_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        280.0   \n",
       "AAAAAACCCGTAGATAGCCT                        565.0   \n",
       "AAAAAAGAAGAAAAAACCAG                         99.0   \n",
       "\n",
       "                      5637-311Cas9_RepB_p6_batch3  59M-311Cas9_RepA_p4_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        489.0                       683.0   \n",
       "AAAAAACCCGTAGATAGCCT                        770.0                      1279.0   \n",
       "AAAAAAGAAGAAAAAACCAG                        102.0                        18.0   \n",
       "\n",
       "                      59M-311Cas9_RepB_p4_batch3  ...  \\\n",
       "AAAAAAATCCAGCAATGCAG                       373.0  ...   \n",
       "AAAAAACCCGTAGATAGCCT                       720.0  ...   \n",
       "AAAAAAGAAGAAAAAACCAG                        76.0  ...   \n",
       "\n",
       "                      YKG1-311Cas9_RepA_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        565.0   \n",
       "AAAAAACCCGTAGATAGCCT                        637.0   \n",
       "AAAAAAGAAGAAAAAACCAG                        171.0   \n",
       "\n",
       "                      YKG1-311Cas9_RepB_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        556.0   \n",
       "AAAAAACCCGTAGATAGCCT                        773.0   \n",
       "AAAAAAGAAGAAAAAACCAG                         57.0   \n",
       "\n",
       "                      ZR-75-1-311Cas9_RepA_p5_batch2  \\\n",
       "AAAAAAATCCAGCAATGCAG                           226.0   \n",
       "AAAAAACCCGTAGATAGCCT                           632.0   \n",
       "AAAAAAGAAGAAAAAACCAG                            44.0   \n",
       "\n",
       "                      ZR-75-1-311Cas9_RepB_p5_batch2  \\\n",
       "AAAAAAATCCAGCAATGCAG                           386.0   \n",
       "AAAAAACCCGTAGATAGCCT                           673.0   \n",
       "AAAAAAGAAGAAAAAACCAG                            39.0   \n",
       "\n",
       "                      pDNA Avana4_010115_1.5Ex_batch0  \\\n",
       "AAAAAAATCCAGCAATGCAG                            305.0   \n",
       "AAAAAACCCGTAGATAGCCT                            398.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             93.0   \n",
       "\n",
       "                      pDNA Avana4_010115_1.5Ex_batch1  \\\n",
       "AAAAAAATCCAGCAATGCAG                            300.0   \n",
       "AAAAAACCCGTAGATAGCCT                            396.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             75.0   \n",
       "\n",
       "                      pDNA Avana4_0101215_0.55Ex_batch1  \\\n",
       "AAAAAAATCCAGCAATGCAG                              190.0   \n",
       "AAAAAACCCGTAGATAGCCT                              254.0   \n",
       "AAAAAAGAAGAAAAAACCAG                               42.0   \n",
       "\n",
       "                      pDNA Avana4_060115_0.55Ex_batch1  \\\n",
       "AAAAAAATCCAGCAATGCAG                             176.0   \n",
       "AAAAAACCCGTAGATAGCCT                             230.0   \n",
       "AAAAAAGAAGAAAAAACCAG                              31.0   \n",
       "\n",
       "                      pDNA Avana4_060115_1.5Ex_batch0  \\\n",
       "AAAAAAATCCAGCAATGCAG                            324.0   \n",
       "AAAAAACCCGTAGATAGCCT                            421.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             74.0   \n",
       "\n",
       "                      pDNA Avana4_060115_1.5Ex_batch1  \n",
       "AAAAAAATCCAGCAATGCAG                            289.0  \n",
       "AAAAAACCCGTAGATAGCCT                            445.0  \n",
       "AAAAAAGAAGAAAAAACCAG                             69.0  \n",
       "\n",
       "[3 rows x 1271 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "readcount.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load CCDS data to decide protein-coding genes\n",
    "ccdsdata = pd.read_table('CCDS.20180614.txt',index_col=None,header=0)\n",
    "ccdsdata = ccdsdata[(ccdsdata['ccds_status'] == 'Public') | (ccdsdata['ccds_status'] == 'Reviewed, update pending')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chromosome</th>\n",
       "      <th>nc_accession</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>ccds_status</th>\n",
       "      <th>cds_strand</th>\n",
       "      <th>cds_from</th>\n",
       "      <th>cds_to</th>\n",
       "      <th>cds_locations</th>\n",
       "      <th>match_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>148398</td>\n",
       "      <td>CCDS2.2</td>\n",
       "      <td>Public</td>\n",
       "      <td>+</td>\n",
       "      <td>925941</td>\n",
       "      <td>944152</td>\n",
       "      <td>[925941-926012, 930154-930335, 931038-931088, ...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>26155</td>\n",
       "      <td>CCDS3.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>-</td>\n",
       "      <td>944693</td>\n",
       "      <td>959239</td>\n",
       "      <td>[944693-944799, 945056-945145, 945517-945652, ...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>84069</td>\n",
       "      <td>CCDS4.1</td>\n",
       "      <td>Public</td>\n",
       "      <td>+</td>\n",
       "      <td>966531</td>\n",
       "      <td>974574</td>\n",
       "      <td>[966531-966613, 966703-966802, 970276-970422, ...</td>\n",
       "      <td>Identical</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #chromosome  nc_accession     gene  gene_id  ccds_id ccds_status cds_strand  \\\n",
       "1           1  NC_000001.11   SAMD11   148398  CCDS2.2      Public          +   \n",
       "2           1  NC_000001.11    NOC2L    26155  CCDS3.1      Public          -   \n",
       "3           1  NC_000001.11  PLEKHN1    84069  CCDS4.1      Public          +   \n",
       "\n",
       "  cds_from  cds_to                                      cds_locations  \\\n",
       "1   925941  944152  [925941-926012, 930154-930335, 931038-931088, ...   \n",
       "2   944693  959239  [944693-944799, 945056-945145, 945517-945652, ...   \n",
       "3   966531  974574  [966531-966613, 966703-966802, 970276-970422, ...   \n",
       "\n",
       "  match_type  \n",
       "1  Identical  \n",
       "2  Identical  \n",
       "3  Identical  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccdsdata.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gene2id_ccds = dict()\n",
    "id2gene_ccds = dict()\n",
    "\n",
    "for i in ccdsdata.index:\n",
    "    if ccdsdata.loc[i]['gene'] in gene2id_ccds and gene2id_ccds[ccdsdata.loc[i]['gene']] != ccdsdata.loc[i]['gene_id']:\n",
    "        print( i) # check dup\n",
    "    gene2id_ccds[ccdsdata.loc[i]['gene']] = ccdsdata.loc[i]['gene_id']\n",
    "    id2gene_ccds[ccdsdata.loc[i]['gene_id']] = ccdsdata.loc[i]['gene']\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ekim8/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  \"\"\"\n",
      "/home/ekim8/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  import sys\n",
      "/home/ekim8/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  if __name__ == '__main__':\n",
      "/home/ekim8/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:10: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  # Remove the CWD from sys.path while we load stuff.\n"
     ]
    }
   ],
   "source": [
    "#map exon for crisprcleanR\n",
    "gene2cds = dict()\n",
    "gene2cdslength = dict()\n",
    "for i in ccdsdata.index:\n",
    "    if ccdsdata.loc[i]['match_type'] == 'Partial':  # no info so skip\n",
    "        continue\n",
    "    if ccdsdata.loc[i]['gene'] not in gene2cds or abs(int(ccdsdata.loc[i]['cds_from'])-int(ccdsdata.loc[i]['cds_to']) ) > gene2cdslength[ccdsdata.loc[i]['gene']]:\n",
    "        # longer gene if same gene\n",
    "        gene2cds[ccdsdata.loc[i]['gene']] = ccdsdata.loc[i]['cds_locations'].replace(\"[\",\"\").replace(\"]\",\"\").replace(\" \",\"\").split(\",\")\n",
    "        gene2cdslength[ccdsdata.loc[i]['gene']] = abs(int(ccdsdata.loc[i]['cds_from'])-int(ccdsdata.loc[i]['cds_to']) )\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ekim8/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  if __name__ == '__main__':\n"
     ]
    }
   ],
   "source": [
    "guidegenemap = pd.read_csv('guide_map.csv',index_col=0,header=0)\n",
    "\n",
    "sgrna2gene = dict()\n",
    "sgrnacount = dict()\n",
    "old2new = dict()\n",
    "finalgenes = set()\n",
    "for i in range(len(guidegenemap.index)):\n",
    "    sgrna = guidegenemap.index[i]\n",
    "    temp = guidegenemap.iloc[i]['gene'].replace(\")\",\"\").split(\" (\")\n",
    "    gene = temp[0]\n",
    "    entrezid = int(temp[1])\n",
    "    if entrezid in id2gene_ccds:\n",
    "        old2new[gene]=gene\n",
    "        if id2gene_ccds[entrezid] != gene:\n",
    "            old2new[gene]=id2gene_ccds[entrezid]\n",
    "            gene = id2gene_ccds[entrezid]\n",
    "        finalgenes.add(gene)\n",
    "        sgrna2gene[sgrna] = gene\n",
    "        if sgrna not in sgrnacount:\n",
    "            sgrnacount[sgrna]=0\n",
    "        sgrnacount[sgrna]+=1\n",
    "\n",
    "#select sgRNA target only one protein-coding gene\n",
    "targetsgrna = dict()\n",
    "for sgrna in sgrnacount:\n",
    "    if sgrnacount[sgrna]==1:\n",
    "        targetsgrna[sgrna] = sgrna2gene[sgrna]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17626"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(finalgenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_alignment</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_alignments</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgrna</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>chr10_112724378_+</td>\n",
       "      <td>SHOC2 (8036)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>chr12_95397391_+</td>\n",
       "      <td>NDUFA12 (55967)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>chr4_76891509_-</td>\n",
       "      <td>SDAD1 (55153)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>chr2_33813513_-</td>\n",
       "      <td>FAM98A (25940)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>chr19_20002409_+</td>\n",
       "      <td>ZNF253 (56242)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGGGCTCCAAAAAGG</td>\n",
       "      <td>chr6_26199835_+</td>\n",
       "      <td>HIST1H2BF (8343)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAACAACACATCAGAGCG</td>\n",
       "      <td>chr14_64688272_-</td>\n",
       "      <td>SYNE2 (23224)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAACTCTGGGAAATGACT</td>\n",
       "      <td>chr1_229440884_+</td>\n",
       "      <td>SPHAR (10638)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAGACAACCTCGCCCTG</td>\n",
       "      <td>chr11_64757251_-</td>\n",
       "      <td>BATF2 (116071)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAGAGCTGTTTGAACAA</td>\n",
       "      <td>chr1_59154718_-</td>\n",
       "      <td>MYSM1 (114803)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       genome_alignment              gene  n_alignments\n",
       "sgrna                                                                  \n",
       "AAAAAAATCCAGCAATGCAG  chr10_112724378_+      SHOC2 (8036)             1\n",
       "AAAAAACCCGTAGATAGCCT   chr12_95397391_+   NDUFA12 (55967)             1\n",
       "AAAAAAGAAGAAAAAACCAG    chr4_76891509_-     SDAD1 (55153)             1\n",
       "AAAAAAGCTCAAGAAGGAGG    chr2_33813513_-    FAM98A (25940)             1\n",
       "AAAAAAGGCTGTAAAAGCGT   chr19_20002409_+    ZNF253 (56242)             1\n",
       "AAAAAAGGGCTCCAAAAAGG    chr6_26199835_+  HIST1H2BF (8343)             1\n",
       "AAAAACAACACATCAGAGCG   chr14_64688272_-     SYNE2 (23224)             1\n",
       "AAAAACTCTGGGAAATGACT   chr1_229440884_+     SPHAR (10638)             1\n",
       "AAAAAGACAACCTCGCCCTG   chr11_64757251_-    BATF2 (116071)             1\n",
       "AAAAAGAGCTGTTTGAACAA    chr1_59154718_-    MYSM1 (114803)             1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guidegenemap.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = guidegenemap.loc[targetsgrna]\n",
    "tempdf = dict()\n",
    "for i in range(len(temp.index)):\n",
    "    gene = temp.iloc[i]['gene'].replace(\")\",\"\").split(\" (\")[0]\n",
    "    entrezid = temp.iloc[i]['gene'].replace(\")\",\"\").split(\" (\")[1]\n",
    "    if gene in old2new and old2new[gene] in finalgenes:\n",
    "        temprow = [ temp.iloc[i]['genome_alignment'] , old2new[gene], entrezid, temp.iloc[i]['n_alignments'] ]\n",
    "        tempdf[temp.index[i]] = temprow\n",
    "        \n",
    "        \n",
    "tempdf = pd.DataFrame.from_dict(tempdf).T.rename(columns={0:\"genome_alignment\",1:\"gene\",2:\"entrezid\",3:\"n_alignments\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGAP4 AMY1A AMY1B AMY1C ANKRD20A1 ANKRD20A2 ANKRD20A3 ANKRD20A4 ARL17A ARL17B BOLA2 BOLA2B C4A C4B CBWD3 CCZ1 CCZ1B CGB1 CGB2 CGB3 CGB5 CGB8 CKMT1A CKMT1B CSH1 CSNK2A3 CTAGE4 CTAGE6 CTAGE8 DDT DDTL DEFA1 DEFA1B DEFA3 DEFB103A DEFB103B DEFB104A DEFB104B DEFB105A DEFB105B DEFB106A DEFB106B DEFB107A DEFB107B DEFB130A DEFB130B EIF3C ELOA3 ELOA3B FAM25C FAM25G FAM72A FAM72B FAM72D FAM86B1 FCGR1A FCGR1B FOXD4L3 FOXD4L6 FRG2 GCOM1 GOLGA6C GOLGA6L9 GOLGA8A GOLGA8B GOLGA8J GOLGA8K GOLGA8O GPR89A GPR89B GRAP GRAPL GTF2IRD2 GTF2IRD2B GYPB HBA1 HBA2 HIST1H4J HIST1H4K HIST2H2AA3 HIST2H2AA4 HIST2H3A HIST2H3C HIST2H4A HIST2H4B IFNA13 KRTAP2-1 KRTAP2-2 KRTAP2-3 KRTAP2-4 KRTAP4-7 KRTAP9-2 KRTAP9-4 KRTAP9-7 LCE1C LCE1F LCE2C LIMS3 LIMS4 LYZL1 LYZL2 MBD3L3 MBD3L4 MBD3L5 MYZAP NBPF15 NBPF4 NBPF6 NOMO3 NPIPA1 NPIPA2 NPIPA3 NPIPA7 NPIPA8 OBP2A OBP2B OR11H12 OR2A1 OR2A42 OR4F16 OR4F17 OR4F29 OR4F3 OR4F4 PGA3 PGA4 PGA5 PLGLB1 PLGLB2 POLR2J POLR2J2 POLR2J3 POTEB POTEB2 POTEJ PPIAL4A PPIAL4C PRAMEF15 PRAMEF5 PRAMEF9 PRR20A PRR20B PRR20C PRR20D PRR20E PRR23D1 PRR23D2 RGPD5 RGPD6 RGPD8 RIMBP3B RIMBP3C SERF1A SERF1B SIGLEC14 SLC35E2A SLX1A SLX1B SMN1 SMN2 SPAG11A SPAG11B SPATA31A5 SPDYE2 SPDYE2B SPRR2A SPRR2B SPRR2D SPRR2E SULT1A3 SULT1A4 TBC1D3 TP53TG3 TP53TG3B TP53TG3C TP53TG3D TRIM49 TRIM49C TRIM74 TVP23C TYW1B UGT2A2 USP17L11 USP17L12 USP17L17 USP17L18 USP17L19 USP17L20 USP17L21 USP17L22 USP17L24 USP17L25 USP17L26 USP17L27 USP17L28 USP17L29 USP17L30 ZNF626 ZNF705D\n"
     ]
    }
   ],
   "source": [
    "print (\" \".join(setdiff1d(list(finalgenes),tempdf['gene'].unique())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save modified guide gene map\n",
    "tempdf.to_csv('guide_gene_map_mod.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add gene symbol to readcount\n",
    "intst = intersect1d(readcount.index,list(targetsgrna.keys()))\n",
    "readcount = readcount.loc[intst]\n",
    "readcount['GENE'] = pd.Series(targetsgrna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>143B-311Cas9_RepA_p6_batch3</th>\n",
       "      <th>2313287-311Cas9_RepA_p5_batch3</th>\n",
       "      <th>2313287-311Cas9_RepB_p5_batch3</th>\n",
       "      <th>253J-311Cas9_RepA_p5_batch3</th>\n",
       "      <th>42-MG-BA-311Cas9_RepA_p6_batch3</th>\n",
       "      <th>42-MG-BA-311Cas9_RepB_p6_batch3</th>\n",
       "      <th>5637-311Cas9_RepA_p6_batch3</th>\n",
       "      <th>5637-311Cas9_RepB_p6_batch3</th>\n",
       "      <th>59M-311Cas9_RepA_p4_batch3</th>\n",
       "      <th>59M-311Cas9_RepB_p4_batch3</th>\n",
       "      <th>...</th>\n",
       "      <th>YKG1-311Cas9_RepB_p6_batch3</th>\n",
       "      <th>ZR-75-1-311Cas9_RepA_p5_batch2</th>\n",
       "      <th>ZR-75-1-311Cas9_RepB_p5_batch2</th>\n",
       "      <th>pDNA Avana4_010115_1.5Ex_batch0</th>\n",
       "      <th>pDNA Avana4_010115_1.5Ex_batch1</th>\n",
       "      <th>pDNA Avana4_0101215_0.55Ex_batch1</th>\n",
       "      <th>pDNA Avana4_060115_0.55Ex_batch1</th>\n",
       "      <th>pDNA Avana4_060115_1.5Ex_batch0</th>\n",
       "      <th>pDNA Avana4_060115_1.5Ex_batch1</th>\n",
       "      <th>GENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>833.0</td>\n",
       "      <td>327.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>546.0</td>\n",
       "      <td>213.0</td>\n",
       "      <td>349.0</td>\n",
       "      <td>280.0</td>\n",
       "      <td>489.0</td>\n",
       "      <td>683.0</td>\n",
       "      <td>373.0</td>\n",
       "      <td>...</td>\n",
       "      <td>556.0</td>\n",
       "      <td>226.0</td>\n",
       "      <td>386.0</td>\n",
       "      <td>305.0</td>\n",
       "      <td>300.0</td>\n",
       "      <td>190.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>1089.0</td>\n",
       "      <td>383.0</td>\n",
       "      <td>391.0</td>\n",
       "      <td>1151.0</td>\n",
       "      <td>736.0</td>\n",
       "      <td>727.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>770.0</td>\n",
       "      <td>1279.0</td>\n",
       "      <td>720.0</td>\n",
       "      <td>...</td>\n",
       "      <td>773.0</td>\n",
       "      <td>632.0</td>\n",
       "      <td>673.0</td>\n",
       "      <td>398.0</td>\n",
       "      <td>396.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>421.0</td>\n",
       "      <td>445.0</td>\n",
       "      <td>NDUFA12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>95.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>...</td>\n",
       "      <td>57.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>SDAD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>236.0</td>\n",
       "      <td>197.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>438.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>252.0</td>\n",
       "      <td>138.0</td>\n",
       "      <td>161.0</td>\n",
       "      <td>378.0</td>\n",
       "      <td>138.0</td>\n",
       "      <td>...</td>\n",
       "      <td>236.0</td>\n",
       "      <td>454.0</td>\n",
       "      <td>214.0</td>\n",
       "      <td>131.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>135.0</td>\n",
       "      <td>FAM98A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>460.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>196.0</td>\n",
       "      <td>251.0</td>\n",
       "      <td>204.0</td>\n",
       "      <td>208.0</td>\n",
       "      <td>338.0</td>\n",
       "      <td>254.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>...</td>\n",
       "      <td>186.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>174.0</td>\n",
       "      <td>193.0</td>\n",
       "      <td>ZNF253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTTGGAGAGATGTACGA</td>\n",
       "      <td>334.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>293.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>369.0</td>\n",
       "      <td>473.0</td>\n",
       "      <td>577.0</td>\n",
       "      <td>...</td>\n",
       "      <td>276.0</td>\n",
       "      <td>314.0</td>\n",
       "      <td>392.0</td>\n",
       "      <td>182.0</td>\n",
       "      <td>199.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>LIPH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTTGGCACAAATACGGG</td>\n",
       "      <td>1112.0</td>\n",
       "      <td>590.0</td>\n",
       "      <td>469.0</td>\n",
       "      <td>888.0</td>\n",
       "      <td>963.0</td>\n",
       "      <td>850.0</td>\n",
       "      <td>1005.0</td>\n",
       "      <td>981.0</td>\n",
       "      <td>1202.0</td>\n",
       "      <td>909.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1062.0</td>\n",
       "      <td>1182.0</td>\n",
       "      <td>1285.0</td>\n",
       "      <td>513.0</td>\n",
       "      <td>503.0</td>\n",
       "      <td>356.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>486.0</td>\n",
       "      <td>RPL10L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTTTCCTCTGAAGTAGT</td>\n",
       "      <td>116.0</td>\n",
       "      <td>371.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>535.0</td>\n",
       "      <td>523.0</td>\n",
       "      <td>283.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>...</td>\n",
       "      <td>447.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>886.0</td>\n",
       "      <td>792.0</td>\n",
       "      <td>530.0</td>\n",
       "      <td>516.0</td>\n",
       "      <td>762.0</td>\n",
       "      <td>771.0</td>\n",
       "      <td>RPS24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTTTCCTCTTCTCGAGG</td>\n",
       "      <td>1065.0</td>\n",
       "      <td>398.0</td>\n",
       "      <td>345.0</td>\n",
       "      <td>1011.0</td>\n",
       "      <td>640.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>565.0</td>\n",
       "      <td>692.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>1439.0</td>\n",
       "      <td>...</td>\n",
       "      <td>836.0</td>\n",
       "      <td>481.0</td>\n",
       "      <td>803.0</td>\n",
       "      <td>281.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>158.0</td>\n",
       "      <td>173.0</td>\n",
       "      <td>268.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>CRISPLD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTTGTTTGTCTCATTCATGG</td>\n",
       "      <td>1748.0</td>\n",
       "      <td>790.0</td>\n",
       "      <td>680.0</td>\n",
       "      <td>1819.0</td>\n",
       "      <td>1048.0</td>\n",
       "      <td>997.0</td>\n",
       "      <td>1440.0</td>\n",
       "      <td>1482.0</td>\n",
       "      <td>1099.0</td>\n",
       "      <td>756.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1931.0</td>\n",
       "      <td>1368.0</td>\n",
       "      <td>974.0</td>\n",
       "      <td>721.0</td>\n",
       "      <td>719.0</td>\n",
       "      <td>485.0</td>\n",
       "      <td>454.0</td>\n",
       "      <td>702.0</td>\n",
       "      <td>666.0</td>\n",
       "      <td>C1orf210</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68456 rows × 1272 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      143B-311Cas9_RepA_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        833.0   \n",
       "AAAAAACCCGTAGATAGCCT                       1089.0   \n",
       "AAAAAAGAAGAAAAAACCAG                         95.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                        236.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                        460.0   \n",
       "...                                           ...   \n",
       "TTTGTTGGAGAGATGTACGA                        334.0   \n",
       "TTTGTTGGCACAAATACGGG                       1112.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                        116.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                       1065.0   \n",
       "TTTGTTTGTCTCATTCATGG                       1748.0   \n",
       "\n",
       "                      2313287-311Cas9_RepA_p5_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                           327.0   \n",
       "AAAAAACCCGTAGATAGCCT                           383.0   \n",
       "AAAAAAGAAGAAAAAACCAG                            32.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                           197.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                           117.0   \n",
       "...                                              ...   \n",
       "TTTGTTGGAGAGATGTACGA                           120.0   \n",
       "TTTGTTGGCACAAATACGGG                           590.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                           371.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                           398.0   \n",
       "TTTGTTTGTCTCATTCATGG                           790.0   \n",
       "\n",
       "                      2313287-311Cas9_RepB_p5_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                           188.0   \n",
       "AAAAAACCCGTAGATAGCCT                           391.0   \n",
       "AAAAAAGAAGAAAAAACCAG                            69.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                            59.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                           125.0   \n",
       "...                                              ...   \n",
       "TTTGTTGGAGAGATGTACGA                           266.0   \n",
       "TTTGTTGGCACAAATACGGG                           469.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                           381.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                           345.0   \n",
       "TTTGTTTGTCTCATTCATGG                           680.0   \n",
       "\n",
       "                      253J-311Cas9_RepA_p5_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        546.0   \n",
       "AAAAAACCCGTAGATAGCCT                       1151.0   \n",
       "AAAAAAGAAGAAAAAACCAG                         72.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                        438.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                        196.0   \n",
       "...                                           ...   \n",
       "TTTGTTGGAGAGATGTACGA                        293.0   \n",
       "TTTGTTGGCACAAATACGGG                        888.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                         99.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                       1011.0   \n",
       "TTTGTTTGTCTCATTCATGG                       1819.0   \n",
       "\n",
       "                      42-MG-BA-311Cas9_RepA_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                            213.0   \n",
       "AAAAAACCCGTAGATAGCCT                            736.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             46.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                            152.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                            251.0   \n",
       "...                                               ...   \n",
       "TTTGTTGGAGAGATGTACGA                            301.0   \n",
       "TTTGTTGGCACAAATACGGG                            963.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                            217.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                            640.0   \n",
       "TTTGTTTGTCTCATTCATGG                           1048.0   \n",
       "\n",
       "                      42-MG-BA-311Cas9_RepB_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                            349.0   \n",
       "AAAAAACCCGTAGATAGCCT                            727.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             60.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                            252.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                            204.0   \n",
       "...                                               ...   \n",
       "TTTGTTGGAGAGATGTACGA                            302.0   \n",
       "TTTGTTGGCACAAATACGGG                            850.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                            535.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                            511.0   \n",
       "TTTGTTTGTCTCATTCATGG                            997.0   \n",
       "\n",
       "                      5637-311Cas9_RepA_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        280.0   \n",
       "AAAAAACCCGTAGATAGCCT                        565.0   \n",
       "AAAAAAGAAGAAAAAACCAG                         99.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                        138.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                        208.0   \n",
       "...                                           ...   \n",
       "TTTGTTGGAGAGATGTACGA                        486.0   \n",
       "TTTGTTGGCACAAATACGGG                       1005.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                        523.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                        565.0   \n",
       "TTTGTTTGTCTCATTCATGG                       1440.0   \n",
       "\n",
       "                      5637-311Cas9_RepB_p6_batch3  59M-311Cas9_RepA_p4_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        489.0                       683.0   \n",
       "AAAAAACCCGTAGATAGCCT                        770.0                      1279.0   \n",
       "AAAAAAGAAGAAAAAACCAG                        102.0                        18.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                        161.0                       378.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                        338.0                       254.0   \n",
       "...                                           ...                         ...   \n",
       "TTTGTTGGAGAGATGTACGA                        369.0                       473.0   \n",
       "TTTGTTGGCACAAATACGGG                        981.0                      1202.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                        283.0                       118.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                        692.0                       301.0   \n",
       "TTTGTTTGTCTCATTCATGG                       1482.0                      1099.0   \n",
       "\n",
       "                      59M-311Cas9_RepB_p4_batch3  ...  \\\n",
       "AAAAAAATCCAGCAATGCAG                       373.0  ...   \n",
       "AAAAAACCCGTAGATAGCCT                       720.0  ...   \n",
       "AAAAAAGAAGAAAAAACCAG                        76.0  ...   \n",
       "AAAAAAGCTCAAGAAGGAGG                       138.0  ...   \n",
       "AAAAAAGGCTGTAAAAGCGT                        85.0  ...   \n",
       "...                                          ...  ...   \n",
       "TTTGTTGGAGAGATGTACGA                       577.0  ...   \n",
       "TTTGTTGGCACAAATACGGG                       909.0  ...   \n",
       "TTTGTTTCCTCTGAAGTAGT                       117.0  ...   \n",
       "TTTGTTTCCTCTTCTCGAGG                      1439.0  ...   \n",
       "TTTGTTTGTCTCATTCATGG                       756.0  ...   \n",
       "\n",
       "                      YKG1-311Cas9_RepB_p6_batch3  \\\n",
       "AAAAAAATCCAGCAATGCAG                        556.0   \n",
       "AAAAAACCCGTAGATAGCCT                        773.0   \n",
       "AAAAAAGAAGAAAAAACCAG                         57.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                        236.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                        186.0   \n",
       "...                                           ...   \n",
       "TTTGTTGGAGAGATGTACGA                        276.0   \n",
       "TTTGTTGGCACAAATACGGG                       1062.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                        447.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                        836.0   \n",
       "TTTGTTTGTCTCATTCATGG                       1931.0   \n",
       "\n",
       "                      ZR-75-1-311Cas9_RepA_p5_batch2  \\\n",
       "AAAAAAATCCAGCAATGCAG                           226.0   \n",
       "AAAAAACCCGTAGATAGCCT                           632.0   \n",
       "AAAAAAGAAGAAAAAACCAG                            44.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                           454.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                           301.0   \n",
       "...                                              ...   \n",
       "TTTGTTGGAGAGATGTACGA                           314.0   \n",
       "TTTGTTGGCACAAATACGGG                          1182.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                           301.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                           481.0   \n",
       "TTTGTTTGTCTCATTCATGG                          1368.0   \n",
       "\n",
       "                      ZR-75-1-311Cas9_RepB_p5_batch2  \\\n",
       "AAAAAAATCCAGCAATGCAG                           386.0   \n",
       "AAAAAACCCGTAGATAGCCT                           673.0   \n",
       "AAAAAAGAAGAAAAAACCAG                            39.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                           214.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                           262.0   \n",
       "...                                              ...   \n",
       "TTTGTTGGAGAGATGTACGA                           392.0   \n",
       "TTTGTTGGCACAAATACGGG                          1285.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                           128.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                           803.0   \n",
       "TTTGTTTGTCTCATTCATGG                           974.0   \n",
       "\n",
       "                      pDNA Avana4_010115_1.5Ex_batch0  \\\n",
       "AAAAAAATCCAGCAATGCAG                            305.0   \n",
       "AAAAAACCCGTAGATAGCCT                            398.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             93.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                            131.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                            174.0   \n",
       "...                                               ...   \n",
       "TTTGTTGGAGAGATGTACGA                            182.0   \n",
       "TTTGTTGGCACAAATACGGG                            513.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                            886.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                            281.0   \n",
       "TTTGTTTGTCTCATTCATGG                            721.0   \n",
       "\n",
       "                      pDNA Avana4_010115_1.5Ex_batch1  \\\n",
       "AAAAAAATCCAGCAATGCAG                            300.0   \n",
       "AAAAAACCCGTAGATAGCCT                            396.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             75.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                            120.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                            153.0   \n",
       "...                                               ...   \n",
       "TTTGTTGGAGAGATGTACGA                            199.0   \n",
       "TTTGTTGGCACAAATACGGG                            503.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                            792.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                            259.0   \n",
       "TTTGTTTGTCTCATTCATGG                            719.0   \n",
       "\n",
       "                      pDNA Avana4_0101215_0.55Ex_batch1  \\\n",
       "AAAAAAATCCAGCAATGCAG                              190.0   \n",
       "AAAAAACCCGTAGATAGCCT                              254.0   \n",
       "AAAAAAGAAGAAAAAACCAG                               42.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                               68.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                              109.0   \n",
       "...                                                 ...   \n",
       "TTTGTTGGAGAGATGTACGA                              140.0   \n",
       "TTTGTTGGCACAAATACGGG                              356.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                              530.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                              158.0   \n",
       "TTTGTTTGTCTCATTCATGG                              485.0   \n",
       "\n",
       "                      pDNA Avana4_060115_0.55Ex_batch1  \\\n",
       "AAAAAAATCCAGCAATGCAG                             176.0   \n",
       "AAAAAACCCGTAGATAGCCT                             230.0   \n",
       "AAAAAAGAAGAAAAAACCAG                              31.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                              90.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                             118.0   \n",
       "...                                                ...   \n",
       "TTTGTTGGAGAGATGTACGA                             118.0   \n",
       "TTTGTTGGCACAAATACGGG                             289.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                             516.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                             173.0   \n",
       "TTTGTTTGTCTCATTCATGG                             454.0   \n",
       "\n",
       "                      pDNA Avana4_060115_1.5Ex_batch0  \\\n",
       "AAAAAAATCCAGCAATGCAG                            324.0   \n",
       "AAAAAACCCGTAGATAGCCT                            421.0   \n",
       "AAAAAAGAAGAAAAAACCAG                             74.0   \n",
       "AAAAAAGCTCAAGAAGGAGG                            135.0   \n",
       "AAAAAAGGCTGTAAAAGCGT                            174.0   \n",
       "...                                               ...   \n",
       "TTTGTTGGAGAGATGTACGA                            172.0   \n",
       "TTTGTTGGCACAAATACGGG                            486.0   \n",
       "TTTGTTTCCTCTGAAGTAGT                            762.0   \n",
       "TTTGTTTCCTCTTCTCGAGG                            268.0   \n",
       "TTTGTTTGTCTCATTCATGG                            702.0   \n",
       "\n",
       "                      pDNA Avana4_060115_1.5Ex_batch1      GENE  \n",
       "AAAAAAATCCAGCAATGCAG                            289.0     SHOC2  \n",
       "AAAAAACCCGTAGATAGCCT                            445.0   NDUFA12  \n",
       "AAAAAAGAAGAAAAAACCAG                             69.0     SDAD1  \n",
       "AAAAAAGCTCAAGAAGGAGG                            135.0    FAM98A  \n",
       "AAAAAAGGCTGTAAAAGCGT                            193.0    ZNF253  \n",
       "...                                               ...       ...  \n",
       "TTTGTTGGAGAGATGTACGA                            141.0      LIPH  \n",
       "TTTGTTGGCACAAATACGGG                            486.0    RPL10L  \n",
       "TTTGTTTCCTCTGAAGTAGT                            771.0     RPS24  \n",
       "TTTGTTTCCTCTTCTCGAGG                            236.0  CRISPLD1  \n",
       "TTTGTTTGTCTCATTCATGG                            666.0  C1orf210  \n",
       "\n",
       "[68456 rows x 1272 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "readcount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# file name info\n",
    "files = pd.DataFrame(columns=['foldchangefile','col','bffile','prfile','batch'])\n",
    "\n",
    "for batch in samples:\n",
    "    cols = ['GENE']\n",
    "    cell2col = dict()\n",
    "    for cell in samples[batch]:\n",
    "        cell2col[cell] = list()\n",
    "        for repid in samples[batch][cell]:\n",
    "            cell2col[cell].append(len(cols))\n",
    "            cols.append(repid)\n",
    "    cont2col = list()\n",
    "    for repid in controls[batch]:   # to avoid column number change after calculating fc, control is the last\n",
    "        cont2col.append(len(cols))\n",
    "        cols.append(repid)\n",
    "    \n",
    "    readcount_batch_file = 'Avana19Q2_readcount.batch'+str(batch)\n",
    "    readcount[cols].index.rename(\"sgRNA\",inplace=True)\n",
    "    readcount[cols].to_csv(readcount_batch_file,sep='\\t')\n",
    "    #print( \"./BAGEL_110.py fc -i \" + readcount_batch_file + \" -o \" + readcount_batch_file + \" -c \" + \",\".join(map(str,cont2col)))\n",
    "    fcfile = readcount_batch_file + \".foldchange\"\n",
    "    for cell in cell2col:\n",
    "        bffile =  fcfile + \".\" + cell + \".bf\"\n",
    "        prfile =  fcfile + \".\" + cell + \".pr\"\n",
    "        files.loc[cell] = [fcfile,\",\".join(map(str,cell2col[cell])), bffile, prfile, batch]\n",
    "        #print( \"./BAGEL_110.py bf -i \" + fcfile + \" -o \" + bffile + \" -c \" + \",\".join(map(str,cell2col[cell])) + \" -e training_ess -n training_noness\")\n",
    "        #print( \"./BAGEL_110.py pr -i \" + bffile + \" -o \" + prfile + \" -e training_ess -n training_noness\")\n",
    "    #print()\n",
    "    #print()\n",
    "    #print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>foldchangefile</th>\n",
       "      <th>col</th>\n",
       "      <th>bffile</th>\n",
       "      <th>prfile</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>ACH-000601</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange</td>\n",
       "      <td>1,2,3,4</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0006...</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0006...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-000496</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange</td>\n",
       "      <td>5,6,7,8</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0004...</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0004...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-000222</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange</td>\n",
       "      <td>9,10,11,12</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0002...</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0002...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-000118</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange</td>\n",
       "      <td>13,14,15,16</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0001...</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0001...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-000791</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange</td>\n",
       "      <td>17,18,19,20</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0007...</td>\n",
       "      <td>Avana19Q2_readcount.batch0.foldchange.ACH-0007...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-001715</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange</td>\n",
       "      <td>740,741</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0017...</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0017...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-000682</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange</td>\n",
       "      <td>742,743</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0006...</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0006...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-000980</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange</td>\n",
       "      <td>744,745</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0009...</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0009...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-001521</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange</td>\n",
       "      <td>746</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0015...</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0015...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ACH-000888</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange</td>\n",
       "      <td>747,748</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0008...</td>\n",
       "      <td>Avana19Q2_readcount.batch3.foldchange.ACH-0008...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>563 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   foldchangefile          col  \\\n",
       "ACH-000601  Avana19Q2_readcount.batch0.foldchange      1,2,3,4   \n",
       "ACH-000496  Avana19Q2_readcount.batch0.foldchange      5,6,7,8   \n",
       "ACH-000222  Avana19Q2_readcount.batch0.foldchange   9,10,11,12   \n",
       "ACH-000118  Avana19Q2_readcount.batch0.foldchange  13,14,15,16   \n",
       "ACH-000791  Avana19Q2_readcount.batch0.foldchange  17,18,19,20   \n",
       "...                                           ...          ...   \n",
       "ACH-001715  Avana19Q2_readcount.batch3.foldchange      740,741   \n",
       "ACH-000682  Avana19Q2_readcount.batch3.foldchange      742,743   \n",
       "ACH-000980  Avana19Q2_readcount.batch3.foldchange      744,745   \n",
       "ACH-001521  Avana19Q2_readcount.batch3.foldchange          746   \n",
       "ACH-000888  Avana19Q2_readcount.batch3.foldchange      747,748   \n",
       "\n",
       "                                                       bffile  \\\n",
       "ACH-000601  Avana19Q2_readcount.batch0.foldchange.ACH-0006...   \n",
       "ACH-000496  Avana19Q2_readcount.batch0.foldchange.ACH-0004...   \n",
       "ACH-000222  Avana19Q2_readcount.batch0.foldchange.ACH-0002...   \n",
       "ACH-000118  Avana19Q2_readcount.batch0.foldchange.ACH-0001...   \n",
       "ACH-000791  Avana19Q2_readcount.batch0.foldchange.ACH-0007...   \n",
       "...                                                       ...   \n",
       "ACH-001715  Avana19Q2_readcount.batch3.foldchange.ACH-0017...   \n",
       "ACH-000682  Avana19Q2_readcount.batch3.foldchange.ACH-0006...   \n",
       "ACH-000980  Avana19Q2_readcount.batch3.foldchange.ACH-0009...   \n",
       "ACH-001521  Avana19Q2_readcount.batch3.foldchange.ACH-0015...   \n",
       "ACH-000888  Avana19Q2_readcount.batch3.foldchange.ACH-0008...   \n",
       "\n",
       "                                                       prfile batch  \n",
       "ACH-000601  Avana19Q2_readcount.batch0.foldchange.ACH-0006...     0  \n",
       "ACH-000496  Avana19Q2_readcount.batch0.foldchange.ACH-0004...     0  \n",
       "ACH-000222  Avana19Q2_readcount.batch0.foldchange.ACH-0002...     0  \n",
       "ACH-000118  Avana19Q2_readcount.batch0.foldchange.ACH-0001...     0  \n",
       "ACH-000791  Avana19Q2_readcount.batch0.foldchange.ACH-0007...     0  \n",
       "...                                                       ...   ...  \n",
       "ACH-001715  Avana19Q2_readcount.batch3.foldchange.ACH-0017...     3  \n",
       "ACH-000682  Avana19Q2_readcount.batch3.foldchange.ACH-0006...     3  \n",
       "ACH-000980  Avana19Q2_readcount.batch3.foldchange.ACH-0009...     3  \n",
       "ACH-001521  Avana19Q2_readcount.batch3.foldchange.ACH-0015...     3  \n",
       "ACH-000888  Avana19Q2_readcount.batch3.foldchange.ACH-0008...     3  \n",
       "\n",
       "[563 rows x 5 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "## parsing Human GENCODE for exon info\n",
    "\n",
    "gene2transcript = dict()\n",
    "transcript2exon = dict()\n",
    "transcript2type = dict()\n",
    "transcript2CDS = dict()\n",
    "transcriptlength = dict()\n",
    "gene2location = dict()\n",
    "with open('gencode.v28lift37.annotation.gtf','r') as fp:\n",
    "    for line in fp:\n",
    "        if line[0] == '#':\n",
    "            continue\n",
    "        linearray = line.rstrip().split(\"\\t\")\n",
    "        #print linearray  # ['chr1', 'HAVANA', 'gene', '3073253', '3074322', '.', '+', '.', 'gene_id \"ENSMUSG00000102693.1\"; gene_type \"TEC\"; gene_name \"RP23-271O17.1\"; level 2; havana_gene \"OTTMUSG00000049935.1\";']\n",
    "        chrom = linearray[0]\n",
    "        featuretype = linearray[2]\n",
    "        start = int(linearray[3])\n",
    "        end = int(linearray[4])\n",
    "        strand = linearray[6]\n",
    "        tags = [x.strip().replace('\"','').split(' ') for x in linearray[8].split(\";\")]\n",
    "        \n",
    "        if featuretype=='exon':\n",
    "            gene=\"\"\n",
    "            transcript=\"\"\n",
    "            #genetype = \"\"\n",
    "            for tag in tags:\n",
    "                if tag[0] == 'gene_name':#'gene_id':\n",
    "                    gene=tag[1].split(\".\")[0]\n",
    "                elif tag[0] == 'transcript_id':\n",
    "                    transcript = tag[1]\n",
    "                elif tag[0] == 'exon_number':\n",
    "                    exon = int(tag[1])\n",
    "                if tag[0] == 'tag' and 'appris_principal' in tag[1]:\n",
    "                    transcript2type[transcript] = tag[1]\n",
    "                \n",
    "                #elif tag[0] == 'tag':\n",
    "            \n",
    "            if gene not in gene2transcript:\n",
    "                gene2transcript[gene]=set()\n",
    "            gene2transcript[gene].add(transcript)\n",
    "            if transcript not in transcript2exon:\n",
    "                transcript2exon[transcript] = dict()\n",
    "                transcriptlength[transcript]=0\n",
    "            transcript2exon[transcript][exon] = (chrom,start,end)\n",
    "            transcriptlength[transcript] += abs(end-start)+1\n",
    "        \n",
    "        \n",
    "        #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# map gene to longest transcript\n",
    "gene2longest = dict()\n",
    "for gene in gene2transcript:\n",
    "    maxt = list(gene2transcript[gene])[0]\n",
    "    maxl = transcriptlength[maxt]\n",
    "    for transcript in gene2transcript[gene]:\n",
    "        if transcript in transcript2type:\n",
    "            if maxl < transcriptlength[transcript]: \n",
    "                maxt = transcript\n",
    "                maxl = transcriptlength[transcript]\n",
    "    gene2longest[gene] = maxt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_exon_num(gene,chrom,location):\n",
    "    result = 0\n",
    "    if gene not in gene2longest:\n",
    "        #print (gene)\n",
    "        return 0\n",
    "    transcript = gene2longest[gene]\n",
    "    if chrom != transcript2exon[transcript][1][0]:\n",
    "        #print (gene, \"Chromosome is different %s != Ref = %s\"%(chrom,transcript2exon[transcript][1][0]))\n",
    "        return 0\n",
    "    \n",
    "    for exon_num in transcript2exon[transcript]:\n",
    "        if transcript2exon[transcript][exon_num][1] <=location and location < transcript2exon[transcript][exon_num][2]:\n",
    "            result = exon_num\n",
    "    '''if result == 0:\n",
    "        print gene, location'''\n",
    "    return result\n",
    "\n",
    "#Generate library map for CRISPRcleanR format\n",
    "guidegenemap_filtered = guidegenemap.loc[intersect1d(list(targetsgrna.keys()),readcount.index)]\n",
    "guidegenemap_CRISPRcleanR = pd.DataFrame(index = intersect1d(list(targetsgrna.keys()),readcount.index),columns=['CODE','GENES','EXONE','CHRM','STRAND','STARTpos','ENDpos'])\n",
    "\n",
    "exones = dict()\n",
    "chroms = dict()\n",
    "strands=dict()\n",
    "starts = dict()\n",
    "ends=dict()\n",
    "for i in range(len(guidegenemap_filtered.index)):\n",
    "    genome_align = guidegenemap_filtered.iloc[i]['genome_alignment']\n",
    "    sgrna = guidegenemap_filtered.index[i]\n",
    "    temp = genome_align.split(\"_\")\n",
    "    chrom = temp[0].replace('chr','')\n",
    "    pos = temp[1]\n",
    "    strand = temp[2]\n",
    "    if strand =='-':\n",
    "        posstart = int(pos)\n",
    "        posend = int(pos) + 19\n",
    "    if strand =='+':\n",
    "        posstart = int(pos) - 19\n",
    "        posend = int(pos)\n",
    "    exones[sgrna] = 'ex' + str(check_exon_num(targetsgrna[sgrna],temp[0],int(pos)))\n",
    "    chroms[sgrna] = chrom\n",
    "    strands[sgrna] = strand\n",
    "    starts[sgrna] = posstart\n",
    "    ends[sgrna] = posend\n",
    "    #guidegenemap_CRISPRcleanR.loc[sgrna] = [targetsgrna[sgrna],targetsgrna[sgrna],'-',chrom,strand,posstart,posend]\n",
    "\n",
    "    \n",
    "guidegenemap_CRISPRcleanR['CODE'] = pd.Series(targetsgrna)\n",
    "guidegenemap_CRISPRcleanR['GENES'] = pd.Series(targetsgrna)\n",
    "guidegenemap_CRISPRcleanR['EXONE'] = pd.Series(exones)\n",
    "guidegenemap_CRISPRcleanR['CHRM'] = pd.Series(chroms)\n",
    "guidegenemap_CRISPRcleanR['STRAND'] = pd.Series(strands)\n",
    "guidegenemap_CRISPRcleanR['STARTpos'] = pd.Series(starts)\n",
    "guidegenemap_CRISPRcleanR['ENDpos'] = pd.Series(starts)\n",
    "guidegenemap_CRISPRcleanR.index.rename(\"sgRNA\",inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CODE</th>\n",
       "      <th>GENES</th>\n",
       "      <th>EXONE</th>\n",
       "      <th>CHRM</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>STARTpos</th>\n",
       "      <th>ENDpos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgRNA</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>SHOC2</td>\n",
       "      <td>SHOC2</td>\n",
       "      <td>ex2</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>112724359</td>\n",
       "      <td>112724359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>NDUFA12</td>\n",
       "      <td>NDUFA12</td>\n",
       "      <td>ex1</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>95397372</td>\n",
       "      <td>95397372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>SDAD1</td>\n",
       "      <td>SDAD1</td>\n",
       "      <td>ex10</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>76891509</td>\n",
       "      <td>76891509</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         CODE    GENES EXONE CHRM STRAND   STARTpos     ENDpos\n",
       "sgRNA                                                                         \n",
       "AAAAAAATCCAGCAATGCAG    SHOC2    SHOC2   ex2   10      +  112724359  112724359\n",
       "AAAAAACCCGTAGATAGCCT  NDUFA12  NDUFA12   ex1   12      +   95397372   95397372\n",
       "AAAAAAGAAGAAAAAACCAG    SDAD1    SDAD1  ex10    4      -   76891509   76891509"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "guidegenemap_CRISPRcleanR.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "guidegenemap_CRISPRcleanR.to_csv('Avana_library_filtered_forCRISPRcleanR',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
